### PR TITLE
Added - check if all files provided as part of the config actually exist

### PR DESCRIPTION
**Pull request by dnLeiting**
This contribution is linked to the following issue: https://github.com/smart-data-lake/smart-data-lake/issues/653

### What changes are included in the pull request?
- This pull requests includes updated to the ConfigLoader class. As part of this change it is checked if all config files included in a sequence exist. If not a ConfigurationException is thrown.
- A test is added to validate the expected behaviour.

### Why are the changes needed?
- Until now no error has been shown if in a sequence of config files a invalid config files has been provided.

### CODE REVIEW

File: sdl-core/src/main/scala/io/smartdatalake/config/ConfigLoader.scala [hunk 1]
```diff
@@ -188,7 +196,7 @@ object ConfigLoader extends SmartDataLakeLogger {
             .partition(_.isDirectory)
           (files ++ directories).flatMap(p => getFilesInBfsOrder(p.getPath)).toSeq
       }
-    } else if (fs.isFile(path) && ConfigFile.canHandleExtension(path)) {
+    } else if (fs.getFileStatus(path).isInstanceOf[FileStatus] && ConfigFile.canHandleExtension(path)) {
```
Comment (zzeekk) [inline sdl-core/src/main/scala/io/smartdatalake/config/ConfigLoader.scala]: Type checks should only be used in rare exceptions in Scala, better:
```suggestion
    } else if (fs.getFileStatus(path).isFile && ConfigFile.canHandleExtension(path)) {
```

File: sdl-core/src/test/scala/io/smartdatalake/config/ConfigLoaderTest.scala [hunk 2]
```diff
@@ -62,6 +62,11 @@ class ConfigLoaderTest extends FlatSpec with Matchers {
     an [ConfigurationException] should be thrownBy ConfigLoader.loadConfigFromFilesystem(Seq("foo/bar"), defaultHadoopConf)
   }
 
+  it must "fail if no configuration files are found for any location provided location in the Sequence" in {
+    a [ConfigurationException] should be thrownBy ConfigLoader.loadConfigFromFilesystem(Seq(getClass.getResource("/config/config.conf").toString, getClass.getResource("/config").toURI.toString + "/error.conf"), defaultHadoopConf)
+    a [ConfigurationException] should be thrownBy ConfigLoader.loadConfigFromFilesystem(Seq(getClass.getResource("/config/config.conf").toString, getClass.getResource("/config").toURI.toString + "/empty"), defaultHadoopConf)
```
Comment (zzeekk) [inline sdl-core/src/test/scala/io/smartdatalake/config/ConfigLoaderTest.scala:71]: It should also fail on an existing directory, where no configuration files are found. Can you please check?

**Review by zzeekk — COMMENTED** [review]
Welcome as SDLB contributor!
Can you please check the additional test case?

File: sdl-core/src/test/scala/io/smartdatalake/config/ConfigLoaderTest.scala (hunk 2 above)
Comment (dnLeiting) [reply sdl-core/src/test/scala/io/smartdatalake/config/ConfigLoaderTest.scala:67]: Added the test.

Comment (dnLeiting) [reply sdl-core/src/test/scala/io/smartdatalake/config/ConfigLoaderTest.scala:71]: Done.

File: sdl-core/src/main/scala/io/smartdatalake/config/ConfigLoader.scala [hunk 3]
```diff
@@ -91,10 +91,22 @@ object ConfigLoader extends SmartDataLakeLogger {
...
+        if (ClasspathConfigFile.canHandleScheme(location))
+          Seq(ClasspathConfigFile(location))
+        else
+          try {
+            var fs = location.getFileSystem(hadoopConf)
+            if(fs.getFileStatus(location).isDirectory & !RemoteIteratorWrapper(fs.listFiles(location, true)).exists(path =>
+                configFileExtensions.contains(path.getPath.toUri.toString.split("\\.").last)
+              ))
+                throw ConfigurationException(s"The provided directory path ${location.toUri.getPath} does not contain any valid config files")
+            getFilesInBfsOrder(location)(location.getFileSystem(hadoopConf))
```
Comment (zzeekk) [inline sdl-core/src/main/scala/io/smartdatalake/config/ConfigLoader.scala]:
```suggestion
            val files = getFilesInBfsOrder(location)(location.getFileSystem(hadoopConf))
            if (files.isEmpty){
              throw ConfigurationException(s"The provided directory path ${location.toUri.getPath} does not contain any valid config files")
            }
            files
```

Comment (dnLeiting) [reply sdl-core/src/main/scala/io/smartdatalake/config/ConfigLoader.scala]: Yes, looks very nice and much shorter. :)

**Review by zzeekk — APPROVED** [review]
:-)

File: sdl-core/src/main/scala/io/smartdatalake/config/ConfigLoader.scala [hunk 4]
```diff
@@ -91,10 +91,21 @@ object ConfigLoader extends SmartDataLakeLogger {
...
+    val configFiles = hadoopPaths.flatMap( location =>
+        if (ClasspathConfigFile.canHandleScheme(location))
+          Seq(ClasspathConfigFile(location))
+        else
+          try {
+            val files = getFilesInBfsOrder(location)(location.getFileSystem(hadoopConf))
+            if (files.isEmpty){
+              throw ConfigurationException(s"The provided directory path ${location.toUri.getPath} does not contain any valid config files")
+            }
+            getFilesInBfsOrder(location)(location.getFileSystem(hadoopConf))
```
Comment (zzeekk) [inline sdl-core/src/main/scala/io/smartdatalake/config/ConfigLoader.scala]: Why do we need to do this twice? Can't we return files?

Comment (dnLeiting) [reply sdl-core/src/main/scala/io/smartdatalake/config/ConfigLoader.scala]: added "files" in line 103 again.

**Pull request merged**